### ISSUE TITLE
Out of Memory Error

### ISSUE THREAD

**Issue by gmcgath** [veraPDF/veraPDF-library#516]
I ran verapdf-gui, version 0.10, on an HP printer manual in PDF form; after about half a minute, it got an Out of Memory Error. The stack dump follows. The file is called LJ1010UG_EN.pdf and is 2431236 bytes long. Operating system: OS X 10.10.5. Running verapdf from the command line produces the same OutOfMemoryError, though of course without any Swing classes being involved.

I was able to fix the problem by adding the parameter -Xmx2048m in a copy of the verapdf script. Perhaps VeraPDF should use this parameter, or some suitable value bigger than Java's conservative default, to minimize the chance of running out of memory.

java -version
java version "1.7.0_79"
Java(TM) SE Runtime Environment (build 1.7.0_79-b15)
Java HotSpot(TM) 64-Bit Server VM (build 24.79-b02, mixed mode)

/Users/gmcgath/Software/VeraPDF/VeraPDF/verapdf-gui ; exit;
07:54:55,116 ERROR AWT-EventQueue-0 CheckerPanel:errorInValidatingOccur:435 - Exception during the validation process
java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: Java heap space
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:188)
	at javax.swing.SwingWorker.get(SwingWorker.java:602)
	at org.verapdf.gui.CheckerPanel.validationEnded(CheckerPanel.java:392)
	at org.verapdf.gui.ValidateWorker.done(ValidateWorker.java:182)
	at javax.swing.SwingWorker$5.run(SwingWorker.java:737)
	at javax.swing.SwingWorker$DoSubmitAccumulativeRunnable.run(SwingWorker.java:832)
	at sun.swing.AccumulativeRunnable.run(AccumulativeRunnable.java:112)
	at javax.swing.SwingWorker$DoSubmitAccumulativeRunnable.actionPerformed(SwingWorker.java:842)
	at javax.swing.Timer.fireActionPerformed(Timer.java:312)
	at javax.swing.Timer$DoPostEvent.run(Timer.java:244)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:312)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:745)
	at java.awt.EventQueue.access$300(EventQueue.java:103)
	at java.awt.EventQueue$3.run(EventQueue.java:706)
	at java.awt.EventQueue$3.run(EventQueue.java:704)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:715)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:242)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:161)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:150)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:146)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:138)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:91)
Caused by: java.lang.OutOfMemoryError: Java heap space
	at org.apache.fontbox.cff.IndexData.initData(IndexData.java:95)
	at org.apache.fontbox.cff.CFFParser.readIndexData(CFFParser.java:163)
	at org.apache.fontbox.cff.CFFParser.parseFont(CFFParser.java:393)
	at org.apache.fontbox.cff.CFFParser.parse(CFFParser.java:115)
	at org.apache.fontbox.ttf.CFFTable.read(CFFTable.java:53)
	at org.apache.fontbox.ttf.TrueTypeFont.readTable(TrueTypeFont.java:361)
	at org.apache.fontbox.ttf.OpenTypeFont.getCFF(OpenTypeFont.java:61)
	at org.apache.pdfbox.pdmodel.font.FileSystemFontProvider.addTrueTypeFontImpl(FileSystemFontProvider.java:260)
	at org.apache.pdfbox.pdmodel.font.FileSystemFontProvider.addTrueTypeCollection(FileSystemFontProvider.java:180)
	at org.apache.pdfbox.pdmodel.font.FileSystemFontProvider.<init>(FileSystemFontProvider.java:155)
	at org.apache.pdfbox.pdmodel.font.FontMapper$DefaultFontProvider.<clinit>(FontMapper.java:80)
	at org.apache.pdfbox.pdmodel.font.FontMapper.getProvider(FontMapper.java:99)
	at org.apache.pdfbox.pdmodel.font.FontMapper.findFont(FontMapper.java:414)
	at org.apache.pdfbox.pdmodel.font.FontMapper.findFontBoxFont(FontMapper.java:383)
	at org.apache.pdfbox.pdmodel.font.FontMapper.getFontBoxFont(FontMapper.java:356)
	at org.apache.pdfbox.pdmodel.font.PDType1Font.<init>(PDType1Font.java:110)
	at org.apache.pdfbox.pdmodel.font.PDType1Font.<clinit>(PDType1Font.java:72)
	at org.apache.pdfbox.pdmodel.font.PDFontFactory.createFont(PDFontFactory.java:62)
	at org.apache.pdfbox.pdmodel.PDResources.getFont(PDResources.java:96)
	at org.verapdf.model.tools.resources.PDInheritableResources.getFont(PDInheritableResources.java:47)
	at org.verapdf.model.factory.operator.OperatorParser.getFontFromResources(OperatorParser.java:489)
	at org.verapdf.model.factory.operator.OperatorParser.parseOperator(OperatorParser.java:250)
	at org.verapdf.model.factory.operator.OperatorFactory.operatorsFromTokens(OperatorFactory.java:49)
	at org.verapdf.model.impl.pb.pd.PBoxPDContentStream.getOperators(PBoxPDContentStream.java:53)
	at org.verapdf.model.impl.pb.pd.PBoxPDContentStream.getLinkedObjects(PBoxPDContentStream.java:41)
	at org.verapdf.pdfa.validators.BaseValidator.addAllLinkedObjects(BaseValidator.java:194)
	at org.verapdf.pdfa.validators.BaseValidator.checkNext(BaseValidator.java:136)
	at org.verapdf.pdfa.validators.BaseValidator.validate(BaseValidator.java:87)
	at org.verapdf.pdfa.validators.BaseValidator.validate(BaseValidator.java:371)
	at org.verapdf.gui.ValidateWorker.runValidator(ValidateWorker.java:171)
	at org.verapdf.gui.ValidateWorker.doInBackground(ValidateWorker.java:95)
	at org.verapdf.gui.ValidateWorker.doInBackground(ValidateWorker.java:46)

**Comment by bdoubrov** [comment]
I found a similar manual on the HP site, and it is over 200 pages long. We'll check if there is some optimization possible, but will follow your suggestion and increase memory both for the CLI and the GUI for now.

**Comment by lwforpres** [comment]
I have replicated this exact Out of Memory Error with a 186-page, 154MB pdf thesis (160,619,570 bytes. Huge, but reality of theses). Times out after about 30 seconds. Running verapdf-gui, version 0.10,  Operating system: OS X 10.11.3.

Will verapdf release an updated version with increased memory able to handle documents of this size?

**Comment by bdoubrov** [comment]
We are still checking what is the best way of increasing memory for both the GUI and the CLI. In particular, right now it is already possible via JAVA_OPTS environment variable. 

I also wonder if you could share any of your test files. Because the HP Lj1010 user manual I managed to find: 
- https://shop.evry.com/pdf/CF1BF285-1DBC-4A4A-BA5C-C07DEF8C79BA.pdf

is validated within ~10 seconds (my  MS Windows laptop) and requires less than 1Gb.

**Comment by freuter** [comment]
Hi,

had the same problem even with "simple" documents. Found that the "fonts" are not cached and that a lot of font instances being created. Which caused an Out Of Memory Exception.

A "quick hack" adding some kind of caching fixed the problem for me (for small and large docs -- and actually improved performance too :-)):
```diff
--- a/model-implementation/src/main/java/org/verapdf/model/factory/operator/OperatorParser.java
+++ b/model-implementation/src/main/java/org/verapdf/model/factory/operator/OperatorParser.java
@@ -56,6 +56,7 @@ import java.io.IOException;
 import java.util.ArrayDeque;
 import java.util.ArrayList;
 import java.util.Deque;
+import java.util.HashMap;
 import java.util.List;
 
 /**
@@ -480,13 +481,24 @@ class OperatorParser {
 		return resources == null ? null : resources.getExtGState(extGState);
 	}
 
+    private static PDInheritableResources _cache_resources=null;
+    private static HashMap<COSName, PDFont> _cache_resources_hash=new HashMap<COSName, PDFont>();
     private static PDFont getFontFromResources(PDInheritableResources resources,
             COSName font) {
 		if (resources == null) {
 			return null;
 		}
         try {
-            return resources.getFont(font);
+        	if (_cache_resources!=resources) {
+        		_cache_resources_hash.clear();
+        		_cache_resources=resources;
+        	}
+        	PDFont ret=_cache_resources_hash.get(font);
+        	if (null==ret) {
+        		ret=resources.getFont(font);
+        		_cache_resources_hash.put(font, ret);
+        	}
+            return ret;
         } catch (IOException e) {
             LOGGER.error(
                     MSG_PROBLEM_OBTAINING_RESOURCE + font + ". "
```

On another note. The master seems to need a fix in the "The file trailer dictionary shall contain the ID keyword whose value shall be File Identifiers as defined in ISO 32000-1:2008, 14.4" rule:
```diff
diff --git a/core/src/main/resources/org/verapdf/pdfa/validation/pdfa-2b.xml b/core/src/main/resources/org/verapdf/pdfa/validation/pdfa-2b.xml
index 8c83f81..61db143 100644
--- a/core/src/main/resources/org/verapdf/pdfa/validation/pdfa-2b.xml
+++ b/core/src/main/resources/org/verapdf/pdfa/validation/pdfa-2b.xml
@@ -549,7 +549,7 @@
         <rule object="CosDocument">
             <id specification="ISO_19005_2" clause="6.1.3" testNumber="1"/>
             <description>The file trailer dictionary shall contain the ID keyword whose value shall be File Identifiers as defined in ISO 32000-1:2008, 14.4</description>
-            <test>lastID</test>
+            <test>null != lastID</test>
             <error>
                 <message>Missing ID in the document trailer</message>
                 <arguments/>
```

**Comment by bdoubrov** [comment]
Thanks a lot for identifying the problem. Surprisingly, we came to the same conclusion (too much memory allocated by font instances) just today in the morning and are working on somewhat similar fix right now. It will be included into the coming 0.12 release planned for the end of March.

The fix for the 6.1.3-1 rule is already in our integration branch and will also be included into 0.12 release:

https://github.com/veraPDF/veraPDF-validation-profiles/blob/integration/PDF_A/2b/6.1%20File%20structure/6.1.3%20File%20trailer/verapdf-profile-6-1-3-t01.xml

**Comment by freuter** [comment]
Great minds think alike :-)

**Comment by bdoubrov** [comment]
Fixed in:

![image](https://cloud.githubusercontent.com/assets/10971979/14022252/0f49fdf8-f1f0-11e5-8eaa-304b214d94e8.png)

Will close the issue after the release

**Comment by bdoubrov** [comment]
Included into the new release (0.12.9): 
http://verapdf.org/2016/04/01/verapdf-0-12-released-alongside-first-version-of-wiki-validation-rules/
